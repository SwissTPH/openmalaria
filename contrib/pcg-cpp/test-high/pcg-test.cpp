/*
 * PCG Random Number Generation for C++
 *
 * Copyright 2014-2017 Melissa O'Neill <oneill@pcg-random.org>,
 *                     and the PCG Project contributors.
 *
 * SPDX-License-Identifier: (Apache-2.0 OR MIT)
 *
 * Licensed under the Apache License, Version 2.0 (provided in
 * LICENSE-APACHE.txt and at http://www.apache.org/licenses/LICENSE-2.0)
 * or under the MIT license (provided in LICENSE-MIT.txt and at
 * http://opensource.org/licenses/MIT), at your option. This file may not
 * be copied, modified, or distributed except according to those terms.
 *
 * Distributed on an "AS IS" BASIS, WITHOUT WARRANTY OF ANY KIND, either
 * express or implied.  See your chosen license for details.
 *
 * For additional information about the PCG random number generation scheme,
 * visit http://www.pcg-random.org/.
 */

/*
 * This file is based on the demo program for the C generation schemes.
 * It shows some basic generation tasks.
 */

#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <cassert>
#include <climits>
#include <iostream>
#include <sstream>
#include <iomanip>
#include <algorithm>
#include <numeric>
#include <random>       // for random_device

#include "pcg_random.hpp"

// This code can be compiled with the preprocessor symbol RNG set to the
// PCG generator you'd like to test.

#ifndef RNG
    #define RNG pcg32
    #define TWO_ARG_INIT 1
#endif

#define STRINGIFY_IMPL(x) #x
#define STRINGIFY(x)      STRINGIFY_IMPL(x)

using namespace std;
using pcg_extras::operator<<;

#if !PCG_EMULATED_128BIT_MATH || !AWKWARD_128BIT_CODE

int main(int argc, char** argv)
{
    // Read command-line options

    int rounds = 5;
    bool nondeterministic_seed = false;

    ++argv;
    --argc;
    if (argc > 0 && strcmp(argv[0], "-r") == 0) {
        nondeterministic_seed = true;
        ++argv;
        --argc;
    }
    if (argc > 0) {
        rounds = atoi(argv[0]);
    }

    /* Many of the generators can be initialized with two arguments; the second
     * one specifies the stream.
     */

#if TWO_ARG_INIT
    RNG rng(42u, 54u);
#else
    RNG rng(42u);
#endif

    if (nondeterministic_seed) {
        // Seed with external entropy from std::random_device (a handy
        // utility provided by pcg_extras).
        rng.seed(pcg_extras::seed_seq_from<std::random_device>());
    }

    constexpr auto bits = sizeof(RNG::result_type) * CHAR_BIT;
    constexpr int how_many_nums = bits <= 8  ? 14
                                : bits <= 16 ? 10
                                :              6;
    constexpr int wrap_nums_at  = bits >  64 ? 2
                                : bits >  32 ? 3
                                :              how_many_nums;

    cout << STRINGIFY(RNG) << ":\n"
    //   << "      -  aka:         " << pcg_extras::printable_typename<RNG>()
    // ^-- we skip this line because the output is long, scary, ugly, and
    //     and varies depending on the platform
         << "      -  result:      " <<  bits << "-bit unsigned int\n"
         << "      -  period:      2^" << RNG::period_pow2();
    if (RNG::streams_pow2() > 0)
         cout << "   (* 2^" << RNG::streams_pow2() << " streams)";
    cout << "\n      -  size:        " << sizeof(RNG) << " bytes\n\n";

    for (int round = 1; round <= rounds; ++round) {
        printf("Round %d:\n", round);

        /* Make some N-bit numbers */
        cout << setw(4) << setfill(' ') << bits << "bit:";
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << endl;

        cout << "  Again:";
        rng.backstep(6);
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << dec << endl;

        /* Toss some coins */
        cout << "  Coins: ";
        for (int i = 0; i < 65; ++i)
            cout << (rng(2) ? "H" : "T");
        cout << endl;

        RNG rng_copy{rng};
        /* Roll some dice */
        printf("  Rolls:");
        for (int i = 0; i < 33; ++i)
            cout << " " << (uint32_t(rng(6)) + 1);
        cout << "\n   -->   rolling dice used "
             << static_cast<size_t>(rng - rng_copy)
             << " random numbers" << endl;

        /* Deal some cards using pcg_extras::shuffle, which follows
         * the algorithm for shuffling that most programmers would expect.
         * (It's unspecified how std::shuffle works.)
         */      
        enum { SUITS = 4, NUMBERS = 13, CARDS = 52 };
        char cards[CARDS];
        iota(begin(cards), end(cards), 0);
        pcg_extras::shuffle(begin(cards), end(cards), rng);

        /* Output the shuffled deck */
        printf("  Cards:");
        static const signed char number[] = {'A', '2', '3', '4', '5', '6', '7',
                                             '8', '9', 'T', 'J', 'Q', 'K'};
        static const signed char suit[] =   {'h', 'c', 'd', 's'};
        int i = 0;
        for (auto card : cards) {
            ++i;
            cout << " " << number[card / SUITS] << suit[card % SUITS];
            if (i % 22 == 0)
                cout << "\n\t";
        }
        
        cout << "\n" << endl;
    }

    // Ensure that the input/output functions work.
    std::stringstream ss;
    ss << rng;
    RNG from_string;
    ss >> from_string;
    assert(from_string == rng);

    return 0;
}

#else //  i.e. PCG_EMULATED_128BIT_MATH && AWKWARD_128BIT_CODE

int main()
{
    // It's not that it *can't* be done, it just requires either C++14-style
    // constexpr or some things not to be declared const.
    cout << "Sorry, " STRINGIFY(RNG) " not supported with emulated 128-bit math"
         << endl;
}

#endif
