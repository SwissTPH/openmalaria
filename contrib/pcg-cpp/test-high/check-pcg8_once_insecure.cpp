#define RNG pcg8_once_insecure
#define TWO_ARG_INIT 1

#include "pcg-test.cpp"

