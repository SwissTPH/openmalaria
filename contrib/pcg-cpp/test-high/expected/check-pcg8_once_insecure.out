pcg8_once_insecure:
      -  result:      8-bit unsigned int
      -  period:      2^8   (* 2^7 streams)
      -  size:        2 bytes

Round 1:
   8bit: 0xea 0x4d 0x8a 0x45 0x6b 0x23 0xcb 0xaa 0xf7 0x63 0xec 0x30 0x39 0xbc
  Again: 0xf7 0x63 0xec 0x30 0x39 0xbc 0x24 0xf4 0x82 0x85 0x8b 0x88 0x28 0x21
  Coins: HTTHTHHTTHTHTHTTTHTHHHTHHHHTHHHHHTTHHTHTTTTTTTTTHTHHHHHHHTHTHTHTT
  Rolls: 5 4 1 5 3 2 2 4 6 3 5 6 3 4 1 3 4 2 4 1 2 5 3 4 6 1 1 6 2 1 5 1 3
   -->   rolling dice used 34 random numbers
  Cards: 2c 6d 3c 4d Th Js 4s 8d 2d 9c 9h 3h 3d Td Tc 7c 5h 8c 5s 4h Qh 2s
	 Ks 6h 7d Jh 5d 8s Qs As Ah 7h Ac 5c 8h Ts 9d Kd Kc Qd Ad 6s 2h Jd
	 Kh 6c Jc 4c Qc 3s 7s 9s

Round 2:
   8bit: 0x98 0xfd 0x95 0x79 0xfc 0xf6 0x73 0x64 0x7d 0xf3 0x8e 0xa6 0x51 0xf1
  Again: 0x7d 0xf3 0x8e 0xa6 0x51 0xf1 0x1f 0x4f 0x62 0x29 0x2e 0x01 0x05 0x72
  Coins: THTTHTTTTHTHTTTHHHHTTHHTTTTTHTHTHTTTHHTHHTHHHTHTTTHTHHTHTHTHTHTHH
  Rolls: 6 6 3 2 4 3 1 4 3 1 5 5 2 2 5 5 4 4 3 2 1 6 4 5 3 6 5 4 5 2 5 1 3
   -->   rolling dice used 34 random numbers
  Cards: Th 3d 5h 9h Kd 7s Ad 7c 2h Qs 6c Ac 7d Jd 4c Tc Ah Ts 2c As 4h 9c
	 6d Kh 9d 5d 3c Td 3s Jc 8h 2d 7h Qc 8c 6h 6s Kc 8s Qh 3h 4d Qd Ks
	 4s 5s 8d 2s Jh 5c Js 9s

Round 3:
   8bit: 0xfa 0xe0 0x07 0xaf 0x5d 0xc5 0x44 0xe8 0x89 0x68 0xa5 0xa2 0x2c 0x81
  Again: 0x89 0x68 0xa5 0xa2 0x2c 0x81 0xe5 0xcf 0x54 0xb5 0x52 0x3e 0xb1 0xd1
  Coins: TTHHTTTTHHHTHTHHTHHTHTTHHHTHHTHHTTTTTTHHHTTHTTHHHTTTHTTHTHHTTTHHH
  Rolls: 1 1 2 5 6 4 5 5 4 2 2 2 3 6 5 6 1 1 4 3 3 4 5 1 3 1 2 3 2 1 1 5 6
   -->   rolling dice used 34 random numbers
  Cards: 6d 3d 4d 8d Ts 9d 4h 5h Js Kc Tc 2c Ah 9h Kd Jc As 5s Kh Ac 2h Ks
	 4s Ad 7c Qs Qh 7s 7h 8c 7d 9s Td 3h 6c 2s Th 5d 4c 3s 9c 5c Jd Jh
	 8s 8h 2d Qc Qd 3c 6h 6s

Round 4:
   8bit: 0x39 0xbc 0x24 0xf4 0x82 0x85 0x8b 0x88 0x28 0x21 0x2d 0x26 0x00 0x61
  Again: 0x28 0x21 0x2d 0x26 0x00 0x61 0xba 0x77 0xbd 0x46 0x8c 0x71 0xdc 0x87
  Coins: THTTTHTHHHTHHHHTHHHHHTTHHTHTTTTTTTTTHTHHHHHHHTHTHTHTTTHTHTTHHHHTT
  Rolls: 6 3 4 1 3 4 2 4 1 2 5 3 4 6 1 1 6 2 1 5 1 3 6 4 6 1 2 3 2 4 3 6 6
   -->   rolling dice used 33 random numbers
  Cards: Ts Qd 7c Tc 8h 2d 7d 3s 2h 6h 6d Qs 5s Ad Qh 8s 3d Js Ac Kh Th 8d
	 Kc 2c 9c 5c Td Jc Jd Kd 8c 3c 6s 4d 3h 9d 4s Ks 6c Qc 9h 5h 7s 2s
	 4c As 7h 9s 5d Ah 4h Jh

Round 5:
   8bit: 0xa6 0x51 0xf1 0x1f 0x4f 0x62 0x29 0x2e 0x01 0x05 0x72 0xe4 0xdb 0xc8
  Again: 0x01 0x05 0x72 0xe4 0xdb 0xc8 0x86 0x4b 0xa0 0x1e 0x9e 0xb4 0xa9 0xb0
  Coins: HTTTHHHHTTHHTTTTTHTHTHTTTHHTHHTHHHTHTTTHTHHTHTHTHTHTHHHHTHHTTHTTT
  Rolls: 5 2 2 5 5 4 4 3 2 1 6 4 5 3 6 5 4 5 2 5 1 3 4 5 4 2 2 5 6 4 6 4 3
   -->   rolling dice used 34 random numbers
  Cards: Qs Jc Kd Th 8c 7d Qc Qd 4c 2c 2s 9h 9s 5s Jd 2h As 8s 5d Td 7c Ts
	 9c Js 7h 5c 4s Ac 6c 6s 9d Tc 3h Ad 3d 8h Jh Qh 4h 4d Kh 2d 5h 6d
	 8d 6h Ks 7s Ah 3c 3s Kc

